tokio-executor/src/task/harness.rs

```diff
@@ -0,0 +1,592 @@
+        }
+
+        if res.is_final_ref() {
+            self.dealloc();
+        }
+    }
+
+    unsafe fn dealloc(self) {
+        // Check causality
+        self.header().executor.with_mut(|_| {});
+        self.header().future_causality.with_mut(|_| {});
+        self.trailer().waker.with_mut(|_| {
+            // we can't check the contents of this cell as it is considered
+            // "uninitialized" data at this point.
+        });
+
+        // Drop task components. This is mostly to properly release loom's data
+        self.header.as_ptr().drop_in_place();
+        self.trailer.as_ptr().drop_in_place();
+
+        raw::dealloc::<T, _>(self.header);
+    }
+
+    // ===== join handle =====
+
+    pub(super) unsafe fn read_output(
+        mut self,
+        dst: *mut Track<super::Result<T::Output>>,
+        state: Snapshot,
+    ) {
+        let src = self.future_or_output as *const Track<super::Result<T::Output>>;
+
+        if state.is_canceled() {
+            dst.write(Track::new(Err(Error::cancelled())));
+        } else {
+            src.copy_to_nonoverlapping(dst, 1);
+        }
+
+        // Before transitioning the state, the waker must be read. It is
+        // possible that, after the transition, we are responsible for dropping
+        // the waker but before the waker can be read from the struct, the
+        // struct is deallocated.
+        let (waker, check) = self.read_join_waker();
+
+        // The operation counts as dropping the join handle
+        let res = self.header().state.complete_join_handle();
+
+        if res.is_released() {
+            // We are responsible for freeing the waker handle
+            check.check();
+            let _ = waker.assume_init();
+        }
+
+        if res.is_final_ref() {
+            self.dealloc();
+        }
+    }
+
+    pub(super) fn store_join_waker(&self, waker: &Waker) -> Snapshot {
+        unsafe {
+            self.trailer().waker.with_mut(|ptr| {
+                (*ptr).as_mut_ptr().replace(Some(waker.clone()));
+            });
+        }
+
+        let res = self.header().state.store_join_waker();
+
+        if res.is_complete() || res.is_canceled() {
+            // Drop the waker here
+            self.trailer()
+                .waker
+                .with_mut(|ptr| unsafe { *(*ptr).as_mut_ptr() = None });
+        }
+
+        res
+    }
+
+    pub(super) fn swap_join_waker(&self, waker: &Waker, prev: Snapshot) -> Snapshot {
+        unsafe {
+            let will_wake = self
+                .trailer()
+                .waker
+                .with(|ptr| (*(*ptr).as_ptr()).as_ref().unwrap().will_wake(waker));
+
+            if will_wake {
+                return prev;
+            }
+
+            // Acquire the lock
+            let state = self.header().state.unset_waker();
+
+            if !state.is_active() {
+                return state;
+            }
+
+            self.store_join_waker(waker)
```

Comment:
Nit/TIOLI: 
```rust
if state.is_active() {
    return self.store_join_waker(waker);
}

state
```
should be equivalent, and removing the `!` might make the logic slightly easier to follow? Up to you.